libvast/vast/system/reader_command.hpp

```diff
@@ -41,23 +45,49 @@ caf::message reader_command(const command& cmd, caf::actor_system& sys,
                             command::argument_iterator first,
                             command::argument_iterator last) {
   VAST_TRACE(VAST_ARG(options), VAST_ARG("args", first, last));
-  auto input = get_or(options, "read", defaults::command::read_path);
-  auto uds = get_or(options, "uds", false);
+  auto uri = caf::get_if<std::string>(&options, "listen");
+  auto file = caf::get_if<std::string>(&options, "read");
+  if (uri && file)
+    return caf::make_message(make_error(ec::invalid_configuration,
+                                        "only one source possible (-r or -l)"));
+  if (!uri && !file) {
+    using inputs = vast::format::reader::inputs;
+    if constexpr (Reader::defaults::input == inputs::inet)
+      uri = Reader::defaults::uri;
+    else
+      file = std::string{Reader::defaults::path};
+  }
   auto global_table_slice = get_or(sys.config(), "vast.table-slice-type",
                                    defaults::system::table_slice_type);
   auto table_slice = get_or(options, "table-slice", global_table_slice);
-  auto factory = vast::factory<table_slice_builder>::get(table_slice);
+  auto factory = vast::factory<vast::table_slice_builder>::get(table_slice);
   if (factory == nullptr)
-    return caf::make_message(make_error(ec::unspecified,
-                                        "unknown table_slice_builder factory"));
-  auto in = detail::make_input_stream(input, uds);
-  if (!in)
-    return caf::make_message(std::move(in.error()));
-  Reader reader{table_slice, std::move(*in)};
-  auto slice_size = get_or(options, "table-slice-size",
-                           defaults::system::table_slice_size);
-  auto src = sys.spawn(source<Reader>, std::move(reader), factory, slice_size);
-  return source_command(cmd, sys, std::move(src), options, first, last);
+    return caf::make_message(
+      make_error(vast::ec::unspecified, "unknown table_slice_builder factory"));
+  if (uri) {
+    endpoint ep;
+    if (!vast::parsers::endpoint(*uri, ep))
+      return caf::make_message(
+        make_error(vast::ec::parse_error, "unable to parse endpoint", *uri));
+    Reader reader{table_slice};
+    auto slice_size = get_or(options, "table-slice-size",
+                             defaults::system::table_slice_size);
+    auto& mm = sys.middleman();
+    auto src = mm.spawn_broker(datagram_source<Reader>, ep.port.number(),
```

Comment:
Before executing this statement, you now need to check whether `ep.port.type() == port::udp`. It should fail for all other types (until we have support for raw TCP sockets).

Reply:
True. I would accept `port::unknown` as well though.

Reply:
And what would `unknown` map to? TCP or UDP?

Reply:
As long as we only have UDP, it is ok to assume that is what the user meant. The correct way to handle inputs without `port_type` would be to ask the `Reader` for the default protocol that it expects and set it to that. Inside the `inet` case, we should have a switch statement with a branch for every implemented type and return error by default. Does that sound good to you?

Reply:
> As long as we only have UDP, it is ok to assume that is what the user meant.

I can see that point. I would just not see a user actually typing `42/?` instead of `42/udp`. That's why I was suggesting to avoid `unknown` in the first place.

> he correct way to handle inputs without `port_type` would be to ask the `Reader` for the default protocol that it expects and set it to that.

Yes. I actually think we should make the `/T` suffix optional in the parser. If no port type is provided in the input, it will be parsed as `unknown`. It will not break anything in the data parser, because the port parser is tried after all other numeric parsers. But it liberates us from using an unconventional endpoint syntax. Users should still be able to write `1.2.3.4:8888` for this to work and `1.2.3.4:8888/udp` if they want/have to be specific.

Reply:
> I can see that point. I would just not see a user actually typing 42/? instead of 42/udp. That's why I was suggesting to avoid unknown in the first place.

The endpoint parser does not rely on the port parser specifically to allow port specified through just the number. See below.

> Yes. I actually think we should make the `/T` suffix optional in the parser. If no port type is provided in the input, it will be parsed as `unknown`. It will not break anything in the data parser, because the port parser is tried after all other numeric parsers.

I tried that yesterday and it actually broke the expression parser, so I decided to just special case the endpoint parser.

Reply:
> The endpoint parser does not rely on the port parser specifically to allow port specified through just the number.

Okay, good.

> I tried that yesterday and it actually broke the expression parser

Eeek, I wonder where my mental model is broken. The expression parser should use the data parser, and the data parser should try all numeric types before trying ports. We'll have to investigate what goes wrong there in a different context.